src/compaction_filter_test.cc

```diff
@@ -0,0 +1,180 @@
+    options_.create_if_missing = true;
+    options_.disable_background_gc = true;
+    options_.disable_auto_compactions = true;
+    options_.compaction_filter =
+        new TestCompactionFilter(options_.min_blob_size);
+
+    DeleteDir(options_.dirname);
+    DeleteDir(dbname_);
+  }
+
+  ~TitanCompactionFilterTest() override {
+    Close();
+    delete options_.compaction_filter;
+    DeleteDir(options_.dirname);
+    DeleteDir(dbname_);
+  }
+
+  static void DeleteDir(const std::string &dirname) {
+    Env *env = Env::Default();
+    std::vector<std::string> filenames;
+    env->GetChildren(dirname, &filenames);
+
+    for (auto &fname : filenames) {
+      env->DeleteFile(dirname + "/" + fname);
+    }
+    env->DeleteDir(dirname);
+  }
+
+  Status Open() {
+    Status s = TitanDB::Open(options_, dbname_, &db_);
+    if (!s.ok()) {
+      return s;
+    }
+    db_impl_ = reinterpret_cast<TitanDBImpl *>(db_);
+    return Status::OK();
+  }
+
+  void Close() {
+    if (!db_)
+      return;
+
+    ASSERT_OK(db_->Close());
+    db_ = nullptr;
+    rocksdb::Options opts;
+    ASSERT_OK(rocksdb::DestroyDB(dbname_, opts));
+  }
+
+  Status Get(const std::string &key, std::string *value) {
+    ReadOptions ropts;
+    return db_->Get(ropts, key, value);
+  }
+
+  Status Put(const std::string &key, const std::string &value) {
+    WriteOptions wopts;
+    return db_->Put(wopts, key, value);
+  }
+
+  std::string GetBigValue() {
+    return std::string(options_.min_blob_size + 1, 'v');
+  }
+
+  void CompactAll() {
+    CompactRangeOptions copts;
+    ASSERT_OK(db_->CompactRange(copts, nullptr, nullptr));
+  }
+
+protected:
+  std::string dbname_;
+  TitanOptions options_;
+  TitanDB *db_{nullptr};
+  TitanDBImpl *db_impl_{nullptr};
+};
+
+TEST_F(TitanCompactionFilterTest, CompactNormalValue) {
+  Status s = Open();
+  ASSERT_OK(s);
+
+  s = Put("mykey", "myvalue");
+  ASSERT_OK(s);
+
+  std::string value;
+  s = Get("mykey", &value);
+  ASSERT_OK(s);
+  ASSERT_EQ(value, "myvalue");
+
+  CompactAll();
+
+  s = Get("mykey", &value);
+  ASSERT_TRUE(s.IsNotFound());
+}
+
+TEST_F(TitanCompactionFilterTest, CompactBlobValue) {
+  Status s = Open();
+  ASSERT_OK(s);
+
+  std::string value = GetBigValue();
+  ASSERT_GT(value.length(), options_.min_blob_size);
+  s = Put("bigkey", value);
+  ASSERT_OK(s);
+
+  std::string value1;
+  s = Get("bigkey", &value1);
+  ASSERT_OK(s);
+  ASSERT_EQ(value1, value);
+
+  CompactAll();
+
+  s = Get("bigkey", &value1);
+  ASSERT_TRUE(s.IsNotFound());
+}
+
+TEST_F(TitanCompactionFilterTest, CompactUpdateValue) {
+  options_.blob_file_discardable_ratio = 0.01;
+  options_.min_blob_size = 1;
+  options_.target_file_size_base = 1;
+  ASSERT_OK(Open());
+
+  ASSERT_OK(db_->Put(WriteOptions(), "update-key", "remain1"));
+  ASSERT_OK(db_->Put(WriteOptions(), "update-another-key", "remain2"));
+  ASSERT_OK(db_->Flush(FlushOptions()));
+  ASSERT_OK(db_->Put(WriteOptions(), "update-key", "remain3"));
```

Comment:
let make it not to `remain3`, which result is expected for later get?

Reply:
I don't quite understand you mean. Can you be more specific?

Reply:
```suggestion
  ASSERT_OK(db_->Put(WriteOptions(), "update-key", "value"));
```
like this, just to make it not removed by compaction filter

Reply:
But now `remain3` will not be removed by compaction filter! Change to `value` will be removed by CF.

Reply:
sorry, it's a typo. I mean making it removed by compaction filter, so we can ensure the stale version is not exposed.